source/MaterialXGenOsl/OslShaderGenerator.cpp

```diff
@@ -493,111 +512,75 @@ void OslShaderGenerator::emitLibraryIncludes(ShaderStage& stage, GenContext& con
-        { Type::INTEGER, ShaderMetadata("widget", Type::STRING, Value::createValueFromStrings("number", Type::STRING->getName())) },
-        { Type::FILENAME, ShaderMetadata("widget", Type::STRING, Value::createValueFromStrings("filename", Type::STRING->getName())) },
-        { Type::BOOLEAN, ShaderMetadata("widget", Type::STRING, Value::createValueFromStrings("checkBox", Type::STRING->getName())) }
-    };
-
-    const std::set<const TypeDesc*> METADATA_TYPE_BLACKLIST =
+    static const std::unordered_map<string, string> GEOMPROP_DEFINITIONS =
     {
-        Type::VECTOR2,  // Custom struct types doesn't support metadata declarations.
-        Type::VECTOR4,  //
-        Type::COLOR4,   //
-        Type::FILENAME, //
-        Type::BSDF      //
+        { "Pobject", "transform(\"object\", P)" },
+        { "Pworld", "P" },
+        { "Nobject", "transform(\"object\", N)" },
+        { "Nworld", "N" },
+        { "Tobject", "transform(\"object\", dPdu)" },
+        { "Tworld", "dPdu" },
+        { "Bobject", "transform(\"object\", dPdv)" },
+        { "Bworld", "dPdv" },
+        { "UV0", "{u,v}" },
+        { "Vworld", "I" }
     };
 
     for (size_t i = 0; i < inputs.size(); ++i)
     {
         const ShaderPort* input = inputs[i];
-
         const string& type = _syntax->getTypeName(input->getType());
-        string value = _syntax->getValue(input, true);
-
-        emitLineBegin(stage);
-        emitString(type + " " + input->getVariable(), stage);
 
-        const string& geomprop = input->getGeomProp();
-        if (!geomprop.empty())
+        if (input->getType() == Type::FILENAME)
         {
-            auto it = GEOMPROP_DEFINITIONS.find(geomprop);
-            if (it != GEOMPROP_DEFINITIONS.end())
-            {
-                value = it->second;
-            }
+            // Shader inputs of type 'filename' (textures) need special handling.
+            // In OSL codegen a 'filename' is translated to the custom type 'textureresource',
+            // which is a struct containing a file string and a colorspace string.
+            // For the published shader interface we here split this into two separate inputs,
+            // which gives a nicer shader interface with widget metadata on each input.
+            
+            ValuePtr value = input->getValue();
+            const string valueStr = value ? value->getValueString() : EMPTY_STRING;
+
+            // Add the file string input
+            emitLineBegin(stage);
+            emitString("string " + input->getVariable() + " = \"" + valueStr + "\"", stage);
+            emitMetadata(input, stage);
+            emitString(",", stage);
+            emitLineEnd(stage, false);
+
+            // Add the colorspace string input
+            emitLineBegin(stage);
+            emitString("string " + input->getVariable() + "_colorspace = \"" + input->getColorSpace() + "\"", stage);
```

Comment:
Hi @niklasharrysson ,
- Won't this always try to add a colorspace input regardless of whether it was defined or not (via `getColorSpace()` 
- It seems also ignore the generation option as to whether or not to emit colorspace ?

- BTW I forget, but how does this get propagated into a "colorspace" input on a OSL `texture()` ? Does the "widget" declaration indicate to add it ?
Thanks.

Reply:
Hi @bernardkwok,

Yes, the parameter for colorspace is always added, but this is the same behavior as before. The change is only to split the parameter from a `textureresource` struct input (that also always hade the colorspace string) into two separate string inputs. The main reason for this is that UI can easier be built from this when we use string inputs instead of a custom struct input. And we can include metadata for these inputs that hints that a widget for colorspace is expected to be used.

So the change is to go from this parameterization:
```
    textureresource  file = {"", ""}
    [[
        string widget = "filename"
    ]],

```
into this new parameterization:
```
    string file = ""
    [[
        string widget = "filename"
    ]],
    string file_colorspace = ""
    [[
        string widget = "colorspace"
    ]],

```
If there is no colorspace set as default value an empty string is set. But the user can of course change this later in the UI built for this shader.

The generation option `GenOption.emitColorTransforms` is used to control if color transform nodes should be inserted or not during generation. I don't think that should affect the shader interface. For example, if an OSL version is used that supports color transformations internally in the `texture` call, then we still want the same shader interface but we want to disable the color transform node insertions to avoid getting double transformations.

The propagation of these parameter values are done same as before, using a `textureresource` struct. A corresponding struct is created from the two string in the main function of the shader and then sent to the image sampler functions. A fix to explicitly set the colorspace string in the `texture` call has been done in PR #1546.

Reply:
Hi @niklasharrysson ,
Thanks for the explanation. It all makes sense now (been a while since I looked at this :)).
I didn't know you had already changed the `texture()` signature, and was mostly thinking of the case of clearing the `colorspace` string if you turn on the `emitColorTransforms` option.